// @mhirba/eslint-config-base/index.js
// reading our prettier configuration
const prettierrc = require('./.prettierrc')

module.exports = {
  env: {
    browser: true,
    es6: true,
    jest: true,
  },
  plugins: ['prettier', 'jest', 'module-resolver'],
  extends: [
    'eslint:recommended',
    'airbnb-base',
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:jest/recommended',
    'plugin:prettier/recommended',
  ],
  parser: 'babel-eslint',
  parserOptions: {
    ecmaVersion: 2018,
  },
  settings: {
    'import/resolver': {
      node: {},
      'babel-module': {},
    },
  },
  globals: {
    process: true,
  },
  rules: {
    // override any config set with .prettierrc files
    'prettier/prettier': ['error', prettierrc],

    // basics
    quotes: ['warn', 'single'],
    semi: ['warn', 'never'],
    camelcase: 'off',
    indent: [
      'warn',
      2,
      {
        SwitchCase: 1,
      },
    ],

    // restrictions
    'no-unused-vars': 'warn',
    'no-param-reassign': 'warn',
    'no-plusplus': 'off',
    'no-return-assign': 'warn',
    'no-console': 'warn',
    'no-eval': 'warn',

    // advanced stuff
    'arrow-parens': ['warn', 'as-needed'],
    'object-curly-newline': 'off',
    'class-methods-use-this': 'off',
    'consistent-return': 'off',
    'comma-dangle': 'off',
    'operator-linebreak': ['warn'],

    // babel stuff
    'module-resolver/use-alias': 2,

    // testing with jest
    'jest/no-disabled-tests': 'warn',
    'jest/no-focused-tests': 'error',
    'jest/no-identical-title': 'error',
    'jest/prefer-to-have-length': 'warn',
    'jest/valid-expect': 'error',
  },
}
